### ISSUE TITLE
Translating add-with-carry to sub-with-borrow

### ISSUE THREAD

**Issue by jadephilipoom** [mit-plv/fiat-crypto#352]
I want to put cases in partial evaluation so that
```
expr_let a := (ident.Z.add_get_carry_concrete s) @@ (x, ident.Z.opp @@ y) in
expr_let b := (ident.Z.add_with_get_carry_concrete s) @@ (ident.snd  @@ a, x', ident.Z.opp @@ y') in
expr_let c := ((Z.add_with_get_carry_concrete s) @@ (ident.snd @@ b, x'', ident.Z.opp @@ y'') in
(ident.fst @@ a, ident.fst @@ b, ident.fst @@ c)
```
becomes
```
expr_let a := (ident.Z.sub_get_borrow s) @@ (x, y) in
expr_let b := (ident.Z.sub_with_get_borrow_concrete s) @@ (ident.snd @@ a, x', y') in
expr_let c := (Z.sub_with_get_borrow_concrete s) @@ (ident.snd @@ b, x'', y'') in
(ident.fst @@ a, ident.fst @@ b, ident.fst @@ c)
```

Now, once upon a time we had this cute idea that we would make `Z.sub_get_borrow s x y` return `- ((x - y) / s` as the borrow, instead of just `(x - y) / s`. So, if `Z.add_get_carry s x (-y) = (a, c)`, then `Z.sub_get_borrow s x y = (a, -c)`. The `Z.opp` on `c` "disappears" because `Z.sub_with_get_borrow` *subtracts* incoming borrows, while `Z.add_with_get_carry` adds them.

Because of this choice, when I try to translate add-opp => sub in the pipeline, I end up with the issue that, to produce an equivalent expression, I have to negate the borrow that comes out of `Z.sub_{with_}get_borrow`. That means I have to destruct the pair, which is really not ideal--especially because the ouput is supposed to be flattened by that point. Ultimately, the problem is that I want to actually not produce an exactly equivalent expression, and I don't have the context to argue that all my transformations together actually come out equivalent.

Thoughts? My current ideas are a) to make a definition of `Z.sub_{with_}get_carry` that actually returns a non-negated borrow and b) to address this at the arithmetic level--basically creating an alternate version of `Rows.sum_rows` that subtracts rows. The latter option would suck, but would work. The former option might fail bounds-checking--at that point we would be passing around a negative number as our "borrow".

**Comment by jadephilipoom** [comment]
Another option could be special wrapper identifiers for carries/borrows, which fails to be interpreted if it appears outside of being consumed by the specific operation that it's meant for. Then we could argue that transforming `ident.carry @@ x` to `ident.borrow @@ -x` is okay, because we can guarantee that the previous was consumed by `add_with_get_carry` and the latter by `sub_with_get_borrow`. This sounds pretty tricky, though.

**Comment by jadephilipoom** [comment]
Okay, I've now had an idea I like better:
1. In partial evaluation, match not just on `(ident.Z.add_get_carry_concrete s) @@ (x, y)` but rather `ident.Let_In @@ ((ident.Z.add_get_carry_concrete s) @@ (x, y), f)`. This lets us control inlining.
2. Make a special identifier called `ident.negate_borrow` or something that negates the second argument of a pair.
3. Do the following transformation steps:

```
expr_let a := (ident.Z.add_get_carry_concrete s) @@ (x, ident.Z.opp @@ y) in
expr_let b := (ident.Z.add_with_get_carry_concrete s) @@ (ident.snd  @@ a, x', ident.Z.opp @@ y') in
expr_let c := ((Z.add_with_get_carry_concrete s) @@ (ident.snd @@ b, x'', ident.Z.opp @@ y'') in
(ident.fst @@ a, ident.fst @@ b, ident.fst @@ c)
```
Transform the first subtraction and inline a `negate_borrow` into the continuation:
```
expr_let a := (ident.Z.sub_get_borrow_concrete s) @@ (x, y) in
expr_let b := (ident.Z.add_with_get_carry_concrete s) @@ (ident.snd  @@ (ident.negate_borrow @@ a), x', ident.Z.opp @@ y') in
expr_let c := ((Z.add_with_get_carry_concrete s) @@ (ident.snd @@ b, x'', ident.Z.opp @@ y'') in
(ident.fst @@ (ident.negate_borrow @@ a), ident.fst @@ b, ident.fst @@ c)
```
```
expr_let a := (ident.Z.sub_get_borrow_concrete s) @@ (x, y) in
expr_let b := (ident.Z.add_with_get_carry_concrete s) @@ (ident.snd  @@ (ident.negate_borrow @@ a), x', ident.Z.opp @@ y') in
expr_let c := ((Z.add_with_get_carry_concrete s) @@ (ident.snd @@ b, x'', ident.Z.opp @@ y'') in
(ident.fst @@ (ident.negate_borrow @@ a), ident.fst @@ b, ident.fst @@ c)
```
When we interpret the arguments to the second add, `ident.snd @@ (ident.negate_borrow @@ x)` becomes `ident.Z.opp @@ (ident.snd @@ x)` and we can use `invert_Z_opp` to detect and reverse it:
```
expr_let a := (ident.Z.sub_get_borrow_concrete s) @@ (x, y) in
expr_let b := (ident.Z.sub_with_get_borrow_concrete s) @@ (ident.snd  @@ a, x', y') in
expr_let c := ((Z.add_with_get_carry_concrete s) @@ (ident.snd @@ b, x'', ident.Z.opp @@ y'') in
(ident.fst @@ (ident.negate_borrow @@ a), ident.fst @@ (ident.negate_borrow @@ b), ident.fst @@ c)
```
Then we do the same thing with the third add, and we transform everything with the form `ident.fst @@ (ident.negate_borrow @@ x)` into `ident.fst @@ x`:
```
expr_let a := (ident.Z.sub_get_borrow_concrete s) @@ (x, y) in
expr_let b := (ident.Z.sub_with_get_borrow_concrete s) @@ (ident.snd  @@ a, x', y') in
expr_let c := ((Z.sub_with_get_borrow_concrete s) @@ (ident.snd @@ b, x'', y'') in
(ident.fst @@ a, ident.fst @@ b, ident.fst @@ c)
```
...and that's the expression we wanted!

**Comment by andres-erbsen** [comment]
Option 4: use the abstract interpretation / forward analysis associated data to mark the borrows as negated instead of substituting an identifier encode that. This would probably push us over the line where we should make a named variatn type for partial information instead of using `sum` in `value_step`, so it would involve paying off some technical debt to get it to look nice. Regarding option 3, I'm worried that adding a LetIn special case with a substitution might turn out to complicate future changes to the partial evaluation. @JasonGross ?

**Comment by jadephilipoom** [comment]
Okay, sounds like option 4 then.

Jason, would you mind taking a stab at this? If you don't have the time, I could also give it a shot, but I would want to schedule in some overlap in our working hours for that.

**Comment by jadephilipoom** [comment]
The `barrett` branch has the code I've been working on--the TODOs near `Print barrett_red256` explain what I want the change to fix.

**Comment by JasonGross** [comment]
Andres and I are currently looking at rewriting the partial evaluation code in a way that will let it not need CPS, and will hopefully be more nicely factored.  I'll make sure that we can support something like option 4 with it.

**Comment by jadephilipoom** [comment]
Any idea when that might be ready? I'm not completely blocked on it, but I'd strongly prefer to be able to do this rewrite by mid-May.

**Comment by JasonGross** [comment]
Currently in progress at https://github.com/JasonGross/fiat-crypto/tree/experiments-no-cps?files=1 .  There's a proof-of-concept handling of neg (I left some comments for you).  I still have to hook it up to the full pipeline (should have that within the next couple of days), and finish filling out the partial evaluation rules (if you want to help out and recode the ones for the operations with carries, etc, that'd be great----and you can comment on if the new setup seems better, or worse, or the same as the old setup).

**Comment by jadephilipoom** [comment]
Sweet. :+1: 

I'd be happy to help recode the carry-using operations and think it's a good test of the new setup. Let me know when it's hooked-up enough to start doing that.

**Comment by jadephilipoom** [comment]
What's the status on this? I'd really like to be able to do this by the end of the week if at all possible.

**Comment by JasonGross** [comment]
The rewriter is coming along, albeit slowly.  Depending on how my first week at Google goes, there's a chance I'll have it finished by the end of the week, but it seems somewhat unlikely.  If you're willing to do somewhat-annoying-but-not-very-hard work that will likely get thrown away in a couple of weeks when the rewriter is actually finished, you can fill in the rewrite rules by hand in https://github.com/mit-plv/fiat-crypto/pull/365 where I have `admit`s, and fix up the notations for your stuff, and then we can merge that before integrating rewrite rules.  In its present state, it builds up to the X25519 stuff, and I'm happy to help if you run into trouble with filling in the `admit`s for the operations you use.

**Comment by jadephilipoom** [comment]
I'm just doing a workaround for now (basically handcoding in the subtractions).

**Comment by JasonGross** [comment]
@jadephilipoom The rewrite rules are working at https://github.com/mit-plv/fiat-crypto/pull/371.  I haven't yet done rewrite rules for the operations that you're using.  I'm in the process of getting your Straightline code to work.

**Comment by JasonGross** [comment]
Closed by #373
